### PR TITLE
Refactor generate.py

### PR DESCRIPTION
**Pull request by samuelgarcia**
This big PR is a strong refactor of generate.py:
  * move InjectTemplatesRecording into generate.py
  * change GeneratorRecording into NoiseGeneratorRecording and improve it
  * move and refactor generate_templates() into generate.py
  * create generate_ground_truth_recording() to generate recording totally **lazy** + sorting 
  * rewrite toy_example() using generate_ground_truth_recording()

TODO add upsampling concept into InjectTemplatesRecording

A major improvement will be to use this for testing api, algos, metrics, visulisation without any disk space and memory use.

### CODE REVIEW

**Comment by samuelgarcia** [comment]
@alejoe91 @h-mayorquin @yger @DradeAW ready to review.
We will need to merge **quickly** because it is a blocker of other important PR like #1941 and #1944 which is needed for better spike_localization.

As a side effect we will be able one to refactor many test base on this to avoid generating and writting to disk many times toy example.

**Comment by samuelgarcia** [comment]
@h-mayorquin : the test you design to checkthat get_traces does not consume more memory than allocated is not passing on macos... It will be hard to debug from my side...

**Comment by alejoe91** [comment]
Can you run pre-commit? Also, tests are failing!

File: src/spikeinterface/core/generate.py [hunk 1]
```diff
@@ -722,19 +666,651 @@ def generate_lazy_recording(
...
+    """
+
+    def __init__(
+        self,
+        sorting: BaseSorting,
+        templates: np.ndarray,
+        nbefore: Union[List[int], int, None] = None,
+        amplitude_factor: Union[List[List[float]], List[float], float, None] = None,
+        parent_recording: Union[BaseRecording, None] = None,
+        num_samples: Union[List[int], None] = None,
```
Comment (DradeAW) [inline src/spikeinterface/core/generate.py:942]: I think `Optional[List[int]]` is better than `Union[List[int], None]`

File: src/spikeinterface/core/generate.py [hunk 2]
```diff
@@ -722,19 +666,651 @@ def generate_lazy_recording(
...
+        Can be None (no scaling).
+        Can be scalar all spikes have the same factor (certainly useless).
+        Can be a vector with same shape of spike_vector of the sorting.
+    parent_recording: BaseRecording | None
+        The recording over which to add the templates.
+        If None, will default to traces containing all 0.
+    num_samples: list[int] | int | None
+        The number of samples in the recording per segment.
+        You can use int for mono-segment objects.
+
```
Comment (DradeAW) [inline src/spikeinterface/core/generate.py:928]: `upsample_vector` description is missing.

**Review by DradeAW — COMMENTED** [review]
I unfortunately don't have the time to do an extensive review, but I think for my part it's ok!

I highlighted some details :)

File: src/spikeinterface/core/generate.py (hunk 1 above)
Comment (samuelgarcia) [reply src/spikeinterface/core/generate.py:942]: Done

File: src/spikeinterface/core/generate.py (hunk 2 above)
Comment (samuelgarcia) [reply src/spikeinterface/core/generate.py:931]: done

**Comment by samuelgarcia** [comment]
> @h-mayorquin : the test you design to checkthat get_traces does not consume more memory than allocated is not passing on macos... It will be hard to debug from my side...

Finally I fixed it.

**Comment by samuelgarcia** [comment]
test are passing and precommit-ci is back

**Comment by alejoe91** [comment]
@samuelgarcia good on my side! Great work!!!

- I chanded `hyperpolarization` to `recovery` (as discussed yesterday)
- Exposed the `decay_power` in generate_templates params (default 1.2-1.8)
- Improved docs and fixed some typos

**Review by alejoe91 — APPROVED** [review]

File: src/spikeinterface/core/generate.py [hunk 3]
```diff
@@ -195,39 +246,53 @@ def synthesize_random_firings(
...
+        if add_shift_shuffle:
+            ## make an interesting autocorrelogram shape
+            # this replace the previous rand_distr2()
+            some = rng.choice(spike_times.size, spike_times.size // 2, replace=False)
+            x = rng.random(some.size)
+            a = refractory_sample
+            b = refractory_sample * 20
+            shift = a + (b - a) * x**2
+            spike_times[some] += shift
+            times0 = times0[(0 <= times0) & (times0 < N)]
```
Comment (h-mayorquin) [inline src/spikeinterface/core/generate.py:285]: The N here is not defined.

Comment (alejoe91) [reply src/spikeinterface/core/generate.py:285]: Thanks @h-mayorquin 

@samuelgarcia can you make a PR to fix it?

Comment (samuelgarcia) [reply src/spikeinterface/core/generate.py:285]: oups. thank for this check. I will fix this

File: src/spikeinterface/core/generate.py [hunk 4]
```diff
@@ -393,75 +455,87 @@ def synthetize_spike_train_bad_isi(duration, baseline_rate, num_violations, viol
...
+    """
 
     def __init__(
         self,
-        durations: List[float],
-        sampling_frequency: float,
         num_channels: int,
+        sampling_frequency: float,
+        durations: List[float],
+        noise_level: float = 5.0,
```
Comment (h-mayorquin) [inline src/spikeinterface/core/generate.py:506]: I think that 1.0 is the sensible default here. Why 5.0?

File: src/spikeinterface/core/generate.py [hunk 5]
```diff
@@ -550,150 +586,59 @@ def get_traces(
...
-        traces = np.multiply(self.amplitudes, traces, dtype=self.dtype, out=traces)
-
-        return traces
+noise_generator_recording = define_function_from_class(
+    source_class=NoiseGeneratorRecording, name="noise_generator_recording"
+)
 
 
-def generate_lazy_recording(
+def generate_recording_by_size(
```
Comment (h-mayorquin) [inline src/spikeinterface/core/generate.py:633]: I wish we went all the way to name this "generate_recording_by_memory_size" to be completly specific.

**Comment by h-mayorquin** [comment]
OK, this new mode of NoiseGenerator is terribly slow for long recordings (one hour).

```
on_the_fly
 127.724 seconds

block_approach:
Execution Time: 5.90863 seconds
```

21 times slower (and that's after the improvements in #1948).


Script:
```python
import time
from spikeinterface.core.generate import NoiseGeneratorRecording
import cProfile


def generate_noise():
    strategy = "on_the_fly"
    strategy = "tile_pregenerated"
    print(strategy)
    recording = NoiseGeneratorRecording(num_channels=32, sampling_frequency=30_000.0, durations=[3600], strategy=strategy)
    for i in range(5):
        print(i)
        x = recording.get_traces()
        del x
    return recording

# Profile the execution
profiler = cProfile.Profile()
profiler.enable()
start_time = time.time()
recording = generate_noise()
end_time = time.time()
profiler.disable()
profiler.print_stats(sort="cumulative")

print(f"Execution Time: {end_time - start_time:.5f} seconds")
print(recording)

```

Complete c profiler.

```
on_the_fly
0
1
2
3
4
         450257 function calls (450251 primitive calls) in 127.724 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.123    0.123  127.724  127.724 dev_profile.py:5(generate_noise)
        5    0.002    0.000  127.600   25.520 baserecording.py:238(get_traces)
        5    9.913    1.983  127.598   25.520 generate.py:585(get_traces)
    18005  115.661    0.006  115.661    0.006 {method 'standard_normal' of 'numpy.random._generator.Generator' objects}
    18007    0.579    0.000    2.024    0.000 {numpy.random._generator.default_rng}
    18007    0.148    0.000    1.046    0.000 contextlib.py:76(inner)
    36014    0.222    0.000    0.508    0.000 _ufunc_config.py:32(seterr)
    18007    0.073    0.000    0.362    0.000 _ufunc_config.py:429(__enter__)
    36012    0.149    0.000    0.304    0.000 <__array_function__ internals>:177(concatenate)
    18007    0.055    0.000    0.273    0.000 _ufunc_config.py:434(__exit__)
    18007    0.240    0.000    0.240    0.000 {function SeedSequence.generate_state at 0x7f1cf2fd5a20}
    36014    0.130    0.000    0.171    0.000 _ufunc_config.py:131(geterr)
    36012    0.109    0.000    0.109    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
    18007    0.052    0.000    0.095    0.000 abc.py:117(__instancecheck__)
    72028    0.091    0.000    0.091    0.000 {built-in method numpy.geterrobj}
    36014    0.064    0.000    0.064    0.000 {built-in method numpy.seterrobj}
    36012    0.047    0.000    0.047    0.000 multiarray.py:148(concatenate)
    18007    0.043    0.000    0.043    0.000 {built-in method _abc._abc_instancecheck}
    18007    0.023    0.000    0.023    0.000 contextlib.py:63(_recreate_cm)
        1    0.000    0.000    0.000    0.000 generate.py:506(__init__)
        1    0.000    0.000    0.000    0.000 generate.py:18(_ensure_seed)
        5    0.000    0.000    0.000    0.000 {built-in method numpy.empty}
        6    0.000    0.000    0.000    0.000 {built-in method builtins.print}
        1    0.000    0.000    0.000    0.000 baserecording.py:34(__init__)
        1    0.000    0.000    0.000    0.000 baserecordingsnippets.py:22(__init__)
        1    0.000    0.000    0.000    0.000 base.py:43(__init__)
        5    0.000    0.000    0.000    0.000 base.py:73(_check_segment_index)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.array}
        1    0.000    0.000    0.000    0.000 _dtype.py:344(_name_get)
      6/3    0.000    0.000    0.000    0.000 abc.py:121(__subclasscheck__)
        2    0.000    0.000    0.000    0.000 {method 'integers' of 'numpy.random._generator.Generator' objects}
      6/3    0.000    0.000    0.000    0.000 {built-in method _abc._abc_subclasscheck}
        5    0.000    0.000    0.000    0.000 baserecording.py:110(get_num_segments)
        1    0.000    0.000    0.000    0.000 _dtype.py:330(_name_includes_bit_suffix)
        5    0.000    0.000    0.000    0.000 base.py:82(ids_to_indices)
        1    0.000    0.000    0.000    0.000 numerictypes.py:356(issubdtype)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.arange}
        1    0.000    0.000    0.000    0.000 generate.py:559(__init__)
        1    0.000    0.000    0.000    0.000 random.py:826(getrandbits)
        1    0.000    0.000    0.000    0.000 generate.py:531(<listcomp>)
        2    0.000    0.000    0.000    0.000 numerictypes.py:282(issubclass_)
        6    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.000    0.000 baserecording.py:121(add_recording_segment)
        1    0.000    0.000    0.000    0.000 baserecording.py:677(__init__)
        2    0.000    0.000    0.000    0.000 {built-in method time.time}
        1    0.000    0.000    0.000    0.000 base.py:111(annotate)
        4    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
        1    0.000    0.000    0.000    0.000 {built-in method posix.urandom}
        1    0.000    0.000    0.000    0.000 {method 'format' of 'str' objects}
        1    0.000    0.000    0.000    0.000 _dtype.py:24(_kind_name)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 base.py:1084(__init__)
        1    0.000    0.000    0.000    0.000 base.py:1091(set_parent_extractor)
        1    0.000    0.000    0.000    0.000 {method 'update' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {built-in method from_bytes}
        1    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}


Execution Time: 127.72357 seconds
NoiseGeneratorRecording: 32 channels - 30.0kHz - 1 segments - 108,000,000 samples 
                         3,600.00s (1.00 hours) - float32 dtype - 12.87 GiB
(spikeinterface_env) @h-laptop$ /home/heberto/miniconda3/envs/spikeinterface_env/bin/python /home/heberto/development/spikeinterface/bin/dev_profile.py
tile_pregenerated
0
1
2
3
4
         150 function calls (146 primitive calls) in 5.909 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.130    0.130    5.909    5.909 dev_profile.py:6(generate_noise)
        5    0.000    0.000    5.771    1.154 baserecording.py:238(get_traces)
        5    5.770    1.154    5.770    1.154 generate.py:585(get_traces)
        1    0.000    0.000    0.008    0.008 generate.py:506(__init__)
        1    0.000    0.000    0.008    0.008 generate.py:559(__init__)
        1    0.007    0.007    0.007    0.007 {method 'standard_normal' of 'numpy.random._generator.Generator' objects}
        3    0.000    0.000    0.000    0.000 {numpy.random._generator.default_rng}
        1    0.000    0.000    0.000    0.000 generate.py:18(_ensure_seed)
        5    0.000    0.000    0.000    0.000 {built-in method numpy.empty}
        3    0.000    0.000    0.000    0.000 contextlib.py:76(inner)
        6    0.000    0.000    0.000    0.000 {built-in method builtins.print}
        6    0.000    0.000    0.000    0.000 _ufunc_config.py:32(seterr)
        3    0.000    0.000    0.000    0.000 _ufunc_config.py:429(__enter__)
        5    0.000    0.000    0.000    0.000 base.py:73(_check_segment_index)
        3    0.000    0.000    0.000    0.000 _ufunc_config.py:434(__exit__)
        1    0.000    0.000    0.000    0.000 _dtype.py:344(_name_get)
        3    0.000    0.000    0.000    0.000 abc.py:117(__instancecheck__)
        3    0.000    0.000    0.000    0.000 {function SeedSequence.generate_state at 0x7fbab8fcda20}
        3    0.000    0.000    0.000    0.000 {built-in method _abc._abc_instancecheck}
        2    0.000    0.000    0.000    0.000 {method 'integers' of 'numpy.random._generator.Generator' objects}
        1    0.000    0.000    0.000    0.000 baserecording.py:34(__init__)
        3    0.000    0.000    0.000    0.000 <__array_function__ internals>:177(concatenate)
        6    0.000    0.000    0.000    0.000 _ufunc_config.py:131(geterr)
        5    0.000    0.000    0.000    0.000 baserecording.py:110(get_num_segments)
      4/2    0.000    0.000    0.000    0.000 abc.py:121(__subclasscheck__)
        1    0.000    0.000    0.000    0.000 _dtype.py:330(_name_includes_bit_suffix)
      4/2    0.000    0.000    0.000    0.000 {built-in method _abc._abc_subclasscheck}
        1    0.000    0.000    0.000    0.000 baserecordingsnippets.py:22(__init__)
        1    0.000    0.000    0.000    0.000 numerictypes.py:356(issubdtype)
       12    0.000    0.000    0.000    0.000 {built-in method numpy.geterrobj}
        1    0.000    0.000    0.000    0.000 baserecording.py:121(add_recording_segment)
        5    0.000    0.000    0.000    0.000 base.py:82(ids_to_indices)
        1    0.000    0.000    0.000    0.000 {built-in method numpy.arange}
        6    0.000    0.000    0.000    0.000 {built-in method numpy.seterrobj}
        1    0.000    0.000    0.000    0.000 random.py:826(getrandbits)
        3    0.000    0.000    0.000    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
        1    0.000    0.000    0.000    0.000 base.py:43(__init__)
        6    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        2    0.000    0.000    0.000    0.000 numerictypes.py:282(issubclass_)
        1    0.000    0.000    0.000    0.000 generate.py:531(<listcomp>)
        4    0.000    0.000    0.000    0.000 {built-in method builtins.issubclass}
        1    0.000    0.000    0.000    0.000 baserecording.py:677(__init__)
        2    0.000    0.000    0.000    0.000 {built-in method time.time}
        1    0.000    0.000    0.000    0.000 {built-in method numpy.array}
        1    0.000    0.000    0.000    0.000 {built-in method posix.urandom}
        3    0.000    0.000    0.000    0.000 multiarray.py:148(concatenate)
        1    0.000    0.000    0.000    0.000 base.py:111(annotate)
        1    0.000    0.000    0.000    0.000 base.py:1091(set_parent_extractor)
        1    0.000    0.000    0.000    0.000 {method 'format' of 'str' objects}
        3    0.000    0.000    0.000    0.000 contextlib.py:63(_recreate_cm)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 _dtype.py:24(_kind_name)
        1    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 {method 'update' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 base.py:1084(__init__)
        1    0.000    0.000    0.000    0.000 {built-in method from_bytes}


Execution Time: 5.90863 seconds
NoiseGeneratorRecording: 32 channels - 30.0kHz - 1 segments - 108,000,000 samples 
                         3,600.00s (1.00 hours) - float32 dtype - 12.87 GiB
```

**Comment by samuelgarcia** [comment]
Thank you @h-mayorquin for having a deeper look on this.
I will open a new PR with some changes and lets discuss over there.

on-the-fly is slower because it rnadomly generate every sample for evry channel!!!
The ratio computaion / real time is not so bad. 127s/3600s
But I think this is good to have the 2 options : on fast and the other with better randomness

**Comment by h-mayorquin** [comment]
Yeah, I actually like your solution a lot. I am hoping that at some point we can think on an optimization that will allow us to have the advantage of your implementation (not requiring a constant memory block) but without the speed performance costs.

But as you say, right now, let's keep the two options as they are useful in two different scenarios.

File: src/spikeinterface/core/tests/test_generate.py [hunk 6]
```diff
@@ -33,134 +59,87 @@ def measure_memory_allocation(measure_in_process: bool = True) -> float:
...
-
-    expected_memory_usage_MiB = memory_after_instanciation_MiB + traces_size_MiB
-    ratio = memory_after_traces_MiB * 1.0 / expected_memory_usage_MiB
-    assertion_msg = (
-        f"Memory after loading traces is {memory_after_traces_MiB} MiB and is {ratio:.2f} times"
-        f"the expected memory usage of {expected_memory_usage_MiB} MiB."
+    after_instanciation_MiB = measure_memory_allocation() / bytes_to_MiB_factor
+    memory_usage_MiB = after_instanciation_MiB - before_instanciation_MiB
+    expected_allocation_MiB = dtype.itemsize * num_channels * noise_block_size / bytes_to_MiB_factor
+    ratio = expected_allocation_MiB / expected_allocation_MiB
```
Comment (h-mayorquin) [inline src/spikeinterface/core/tests/test_generate.py:90]: @samuelgarcia You modified this test so it always passes? : P

Comment (samuelgarcia) [reply src/spikeinterface/core/tests/test_generate.py:90]: hé hé. I see it now.
this is some kind of a mistake.

Comment (h-mayorquin) [reply src/spikeinterface/core/tests/test_generate.py:90]: Let's fix it later : D

**Pull request merged**